backend.py

```diff
@@ -0,0 +1,502 @@
+            hbar (float): the value of :math:`\hbar` in the commutation relation :math:`[\x,\p]=i\hbar`.
+            rtol (float): the relative tolerance parameter used in `np.allclose`
+            atol (float): the absolute tolerance parameter used in `np.allclose
+            `
+            Returns:
+               np.float64 : probability of detection pattern
+        """
+
+        if np.allclose(mu, 0, atol=atol, rtol=rtol):
+            # no displacement
+            n_modes = cov.shape[0] // 2
+            Q = Qmat(cov, hbar)
+            O = Xmat(n_modes) @ Amat(cov, hbar=hbar)
+            rpt2 = np.concatenate((det_pattern, det_pattern))
+            Os = self.reduction(O, rpt2)
+            return self.tor(Os) / np.sqrt(np.linalg.det(Q))
+        det_pattern = np.asarray(det_pattern).astype(np.int8)
+        return self.threshold_detection_prob_displacement(mu, cov, det_pattern, hbar)
+
+    def measure_threshold(self, modes, shots=1, select=None, **kwargs):
+        if shots != 1:
+            if select is not None:
+                raise NotImplementedError(
+                    "Gaussian backend currently does not support " "postselection"
+                )
+            warnings.warn(
+                "Cannot simulate non-Gaussian states. "
+                "Conditional state after Threshold measurement has not been updated."
+            )
+
+        mu = self.circuit.mean
+        cov = self.circuit.scovmatxp()
+        # check we are sampling from a gaussian state with zero mean
+        if not allclose(mu, zeros_like(mu)):
+            """For gaussian state with non-zero mean"""
+            return self.threshold_detection_prob_displacement(mu, cov, self.det_pattern, self.hbar)
+
+        x_idxs = array(modes)
+        p_idxs = x_idxs + len(mu)
+        modes_idxs = concatenate([x_idxs, p_idxs])
+        reduced_cov = cov[ix_(modes_idxs, modes_idxs)]
+        samples = torontonian_sample_state(reduced_cov, shots)
+
+        return samples
+
+    def state(self, modes=None, **kwargs):
+        """Returns the state of the quantum simulation.
+
+        See :meth:`.BaseBackend.state`.
+
+        Returns:
+            GaussianState: state description
+        """
+        m = self.circuit.scovmat()
+        r = self.circuit.smean()
+
+        if modes is None:
+            modes = list(range(len(self.get_modes())))
+
+        listmodes = list(concatenate((2 * array(modes), 2 * array(modes) + 1)))
+        covmat = empty((2 * len(modes), 2 * len(modes)))
+        means = r[listmodes]
+
+        for i, ii in enumerate(listmodes):
+            for j, jj in enumerate(listmodes):
+                covmat[i, j] = m[ii, jj]
+
+        means *= sqrt(2 * self.circuit.hbar) / 2
+        covmat *= self.circuit.hbar / 2
+
+        mode_names = ["q[{}]".format(i) for i in array(self.get_modes())[modes]]
+        return BaseGaussianState((means, covmat), len(modes), mode_names=mode_names)
```

Comment:
@Hannibalcarthaga : before reviewing, can I suggest you import all the new functions in this file from `thewalrus` instead of copy pasting them here.

Reply:
@nquesada sorry to bother you again, could you please elaborate on the new functions you are referring to? Also import means commit the changes to this file right? These three functions:                                                                                                                        from .libwalrus import torontonian_complex as tor_complex, from .libwalrus import torontonian_real as tor_real and from . import reduction, are producing an Unable to import error.

Reply:
The function I meant are `tor`, `numba_ix`, `nb_block`, `Qmat_numba`, `powerset`, `threshold_detection_prob_displacement`, `threshold_detection_prob`.

Reply:
Hi @Hannibalcarthaga, 
If you're having trouble importing certain functions or submodules, perhaps you could show the exact code you're using and the full traceback (the full message accompanying the "Unable to import error" that you mentioned). This will help us see what the issue is and hopefully steer you in the right direction.

Reply:
Hi, @co9olguy, I imported the functions that @nquesada mentioned using `from thewalrus._torontonian import...`. Now the definitions are not required(i.e. should I remove them?) and only the `measure_threshold` should be defined. Am I approaching this issue correctly?

Reply:
Hi @Hannibalcarthaga --- Yes you should remove every line code that you can import from other libraries.

Reply:
Hey, @nquesada instead of using `from thewalrus._torontonian import...`, is it possible to use `import thewalrus._torontonian` directly? Also I have made the necessary changes as you suggested,  shall I commit them?

Reply:
We prefer if you do `from library import functions you need`